[Emmanuel Er] iP

emmannyyy opened the pull request:
### Monique Personal ChatBot Assistant

> "CS2103 work load quite rabz.. why not use this chatbot to lighten your workload?” – ME LOL

Monique frees your mind of having to remember things you need to do. It's,

- Text Based
- Easy to learn
- ~~Fast~~ Blazing fast to use

All you need to do is,


download it from [here](https://github.com/emmannyyy/ip/releases/download/A-Jar/Monique.jar).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉   

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines 
- [x] Managing Events
- [x] Finding tasks
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

emmannyyy committed:
Add level 0

emmannyyy committed:
Add level 0

emmannyyy committed:
Merge branch 'master' of https://github.com/emmannyyy/ip

# Conflicts:
#	src/main/java/Monique.java

emmannyyy committed:
add level 1

emmannyyy committed:
Merge branch 'master' of https://github.com/emmannyyy/ip

# Conflicts:
#	src/main/java/Monique.java

emmannyyy committed:
Add level 1 and fixed tags

emmannyyy committed:
Added level-2

emmannyyy committed:
Added Task and Todo classes for level-3

emmannyyy committed:
Added level-5

emmannyyy committed:
Edited Auto Testing Script for A-TextUiTesting

emmannyyy committed:
added exceptions and error handling to level-5

emmannyyy committed:
Added Collections and Level-6

emmannyyy committed:
Include Database Storing and Loading as per Level-7

Previously did not allow for persistent data storage.

Does not fulfill user needs, and has not hit Level-7

This change updates the chatbot to allow for persistent data.

emmannyyy committed:
Merge branch 'heads/level-7' of https://github.com/emmannyyy/ip

* 'heads/level-7' of https://github.com/emmannyyy/ip:
  Include Database Storing and Loading as per Level-7

emmannyyy committed:
Include Date Functionality

Previously, all tasks that had dates (i.e: Deadline and Event) were simply using strings to format dates.

Since this would not allow for more functionality for later features, and is also prone to errors, this is suboptimal and should be fixed to allow for date time functionality.

Let's update our product to allow for this functionality, and to hit Level-8

emmannyyy committed:
Merge branch 'branch-Level-8'

emmannyyy committed:
implement more OOP

Main class did not make use of Abstraction principles in OOP.

This project has a lot of repeated code, making it inefficient and difficult to maintain.

This update will allow for good abstraction and use of OOP concepts

emmannyyy committed:
remove tasks.txt (temporary file), update gitignore

emmannyyy committed:
Merge tag 'A-MoreOOP'

* tag 'A-MoreOOP':
  remove tasks.txt (temporary file), update gitignore
  implement more OOP

emmannyyy committed:
package classes

emmannyyy committed:
Merge branch 'A-Packages'

* A-Packages:
  package classes

emmannyyy committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

* origin/add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

emmannyyy committed:
update gradle build to fit repository structure

emmannyyy committed:
Merge branch 'A-Gradle'

* A-Gradle:
  update gradle build to fit repository structure
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

emmannyyy committed:
add Junit tests to ensure working order of at least 2 functions from 2 different classes

emmannyyy committed:
allow for Jar builds through gradle

emmannyyy committed:
add find functionality for the chatbot

emmannyyy committed:
add javadoc comments and improve formatting of certain classes

emmannyyy committed:
update code to meet coding standard as specified.

emmannyyy committed:
Merge tag 'A-CodingStandard'

emmannyyy committed:
Merge tag 'A-JavaDoc'

# Conflicts:
#	src/main/java/monique/GuideText.java
#	src/main/java/monique/command/AddCommand.java
#	src/main/java/monique/command/ByeCommand.java
#	src/main/java/monique/command/DeleteCommand.java
#	src/main/java/monique/command/GuideCommand.java
#	src/main/java/monique/command/ListCommand.java
#	src/main/java/monique/command/MarkCommand.java
#	src/main/java/monique/command/UnmarkCommand.java
#	src/main/java/monique/exception/DeleteException.java
#	src/main/java/monique/exception/IllegalDateFormatException.java
#	src/main/java/monique/exception/MarkException.java
#	src/main/java/monique/exception/ParseException.java
#	src/main/java/monique/exception/UnknownCommandException.java
#	src/main/java/monique/parser/Parser.java
#	src/main/java/monique/task/Deadline.java
#	src/main/java/monique/task/Event.java
#	src/main/java/monique/task/Task.java
#	src/main/java/monique/task/ToDo.java
#	src/main/java/monique/tasklist/TaskList.java
#	src/main/java/monique/ui/Ui.java

emmannyyy committed:
Merge tag 'A-JavaDoc'

# Conflicts:
#	src/main/java/monique/GuideText.java
#	src/main/java/monique/command/AddCommand.java
#	src/main/java/monique/command/ByeCommand.java
#	src/main/java/monique/command/DeleteCommand.java
#	src/main/java/monique/command/GuideCommand.java
#	src/main/java/monique/command/ListCommand.java
#	src/main/java/monique/command/MarkCommand.java
#	src/main/java/monique/command/UnmarkCommand.java
#	src/main/java/monique/exception/DeleteException.java
#	src/main/java/monique/exception/IllegalDateFormatException.java
#	src/main/java/monique/exception/MarkException.java
#	src/main/java/monique/exception/ParseException.java
#	src/main/java/monique/exception/UnknownCommandException.java
#	src/main/java/monique/parser/Parser.java
#	src/main/java/monique/task/Deadline.java
#	src/main/java/monique/task/Event.java
#	src/main/java/monique/task/Task.java
#	src/main/java/monique/task/ToDo.java
#	src/main/java/monique/tasklist/TaskList.java
#	src/main/java/monique/ui/Ui.java

emmannyyy committed:
deconflict merge

emmannyyy committed:
update methods to match coding standard and change some variables from public static to private final

emmannyyy committed:
add gui for level 10

emmannyyy committed:
update all files to include javadoc comments and remove unused methods

emmannyyy committed:
Merge tag 'Level-10'

emmannyyy committed:
include checkstyle formatting

emmannyyy committed:
Merge tag 'A-CheckStyle'

jarrellc1 commented on src/main/java/monique/command/DeleteCommand.java:
```diff
@@ -10,23 +10,23 @@ public class DeleteCommand extends Command {
             throw new DeleteException();
         }
         Task deletedTask = tasks.getTask(this.taskNum);
         tasks.deleteTask(this.taskNum);
```
Nice job on ensuring the task deletion logic is handled properly. The boundary checks before deleting a task are clear and avoid potential issues with invalid task numbers.

jarrellc1 commented on src/main/java/monique/parser/Parser.java:
```diff
@@ -20,136 +20,135 @@
+                break;
+            }
+            case "list": {
+                command = new ListCommand();
```
Switch case format follows CS2103T format. Good job.
In the mark case, you've included the comment // minus one bc 0-based indexing, which is good for readability. However, to make the code even clearer, you might consider defining a constant, such as INDEX_OFFSET = 1, to avoid the use of hardcoded values.

emmannyyy replied:
Thanks, I've fixed it!

jarrellc1 commented on src/main/java/monique/parser/Parser.java:
```diff
@@ -20,136 +20,135 @@
+                    if (!hasSecondWord) {
+                        throw new ParseException();
                     }
+                    int taskNum = Integer.parseInt(fullCommand.split("delete ")[1]) - 1;
```
same here, the -1 due to line and array difference can be in the form of a constant to improve readability

jarrellc1 commented on src/main/java/monique/parser/Parser.java:
```diff
@@ -20,136 +20,135 @@
+                    if (toSplit.length != 2) {
+                        throw new ParseException();
                     }
+                    String[] commandAndDescription = fromSplit[0].trim().split(" ", 2);
```
Excellent work handling the parsing of complex user inputs, especially for event commands with /from and /to delimiters. The use of clear exception handling to manage parsing issues is a strong approach that will help prevent user errors from crashing the program.

jarrellc1 reviewed (commented):
Overall, your code has followed the coding standard as stated in the textbook, good job.

emmannyyy committed:
use varargs for find functionality, allowing the user to find tasks with multiple search phrases

emmannyyy committed:
Merge branch 'A-Varargs'

Andrew22Teoh commented on src/main/java/monique/exception/DeleteException.java:
```diff
@@ -0,0 +1,26 @@
+     * Constructs a new <code>DeleteException</code> with a default detail message.
+     * The message indicates that the item number is not allowed.
+     */
+    public DeleteException() {
```
Would it be better to initialise DeleteExceptions with custom error messages, in case future improvements lead to more than 1 reason for a DeleteException to be thrown?

Andrew22Teoh commented on src/main/java/monique/parser/Parser.java:
```diff
@@ -0,0 +1,169 @@
+                    }
+                    int taskNum = Integer.parseInt(fullCommand.split("mark ")[1]) -1;
+                    command = new MarkCommand(taskNum);
+                    break;
```
I think this break is unnecessary because of the finally break?

emmannyyy replied:
Thanks! I've fixed it :)

Andrew22Teoh commented on src/main/java/monique/parser/Parser.java:
```diff
@@ -0,0 +1,169 @@
+                    command = new AddCommand(taskToAdd);
+                    break;
+                } catch (ParseException pe) {
+                    pe.advice();
```
What is the String output of this pe.advice() being assigned to?

Andrew22Teoh commented on src/main/java/monique/storage/Storage.java:
```diff
@@ -0,0 +1,107 @@
+ */
+public class Storage {
+    private final String filePath;
+    private final File dbFile;
```
Naming this variable dbFile might be slightly unfamiliar for some users who don't frequently see "db" in place of "database", could I suggest databaseFile or saveFile instead?

Andrew22Teoh commented on src/main/java/monique/storage/Storage.java:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @return The list of tasks loaded from the storage file.
+     */
+    public ArrayList<Task> loadData() {
```
I notice many short variable names present in this method, would it be feasible to convert them to more interpretable variable names?
Eg. ObjectInputStream could be named objectReader instead of ois.

Andrew22Teoh commented on src/main/java/monique/task/Deadline.java:
```diff
@@ -0,0 +1,126 @@
+ */
+public class Deadline extends Task {
+
+    private static final String FORMATSTRING = "[D][%s] %s (by: %s)";
```
Would be good to write this in screaming snake case, as FORMAT_STRING! 😄

emmannyyy replied:
Thanks! I've fixed it :)

Andrew22Teoh reviewed (commented):
Nicely written program! I've left some comments for your consideration that might improve clarity for some readers.

emmannyyy committed:
Update files to meet code quality

The current code is mostly okay, but still violates some of the standards of the coding quality as specified in the course.

Since we are given more thorough coding standards, the previous coding quality is subpar and must be updated.

Let's update the code to meet the coding standards

emmannyyy committed:
Update code to use assertions

The current code does not use assertions, but only does if-else statements for checking of preconditions.

Since we are taught better coding practices involving the assertions, this old method of precondition checking is subpar.

Let's use java assertions to improve code quality and safety.

emmannyyy committed:
Update code to use streams

The current code makes use of for loops instead of streams, which may result in less safe handling of operations.

Since we are taught streams, this use of for loops may provide less safe results as compared to streams.

Let's update the code to use streams instead of for loops, to improve code quality and safety.

emmannyyy committed:
Merge pull request #1 from emmannyyy/A-CodeQuality

Update files to meet code quality

emmannyyy committed:
Merge pull request #2 from emmannyyy/A-Assertions

Update code to use assertions

emmannyyy committed:
Merge pull request #3 from emmannyyy/A-Streams

Update code to use streams

emmannyyy committed:
Merge branches from upstream repo to current master branch locally

emmannyyy committed:
Set up CI using yml file

emmannyyy committed:
Update code to handle time for Deadline and Event classes

Currently, the chatbot only handles dates in very specific formats which are not natural to users.

Since we want users to find this chatbot intuitive and user friendly, it would be subpar to not have this feature.

Let's update the Parser class, Event class and Deadline class to allow for usage of more natural dates.

emmannyyy committed:
Merge branch 'C-NaturalDates'

emmannyyy committed:
Update handling of time and improve abstraction

emmannyyy committed:
Update gradle yml file to fix workflow errors

emmannyyy committed:
Update code to be compliant with coding standard

emmannyyy committed:
Update Guide Text and improve parser class to meet coding standard

emmannyyy committed:
Update guide text to fit checkstyle

emmannyyy committed:
Update Junit tests to have @BeforeAll annotation to set a consistent time locale

emmannyyy committed:
Update code to meet single level of abstraction principle

emmannyyy committed:
Update code to implement a better GUI

emmannyyy committed:
Update code to have add more of a personality

emmannyyy committed:
Update code to include more error handling for event command

emmannyyy committed:
Update code to include more rigorous testing of event command

emmannyyy committed:
Upload Ui picture

emmannyyy committed:
Update readme file to add user guide

emmannyyy committed:
Update code to enforce time locale as sg time

emmannyyy committed:
Update markdown file to include link to jar file

emmannyyy committed:
Update readme file and unmark command

emmannyyy committed:
Update global readme file

emmannyyy committed:
Update global readme file

emmannyyy committed:
Update code to fix bug related to unmark command
